### PR TITLE
Max pool grad grad

### PR DESCRIPTION
**Pull request by aam-at**
Adds gpu and cpu kernels for max pool grad. Supports NHWC and NCHW layouts for gpu kernel and only NHWC for cpu kernel. Additionally registers gradient for python `MaxPoolGradGrad` which is just `MaxPoolGrad`. Registers gradient for python `AvgPoolGrad` which is `AvgPool` evaluated on `top_diff`.

Continuation of #6299

### CODE REVIEW

**Comment by aam-at** [comment]
I signed it!

**Comment by aam-at** [comment]
I did some refactoring for `MaxPoolingOp` which now uses functors and templates and registers kernels using `TF_CALL_*`. Since currently there more kernels for gpu then cpu, functors are not generalized to devices. Also adding `double` required a minor tweak to `stream_executor` (merely duplicating methods for `double`). I also added python tests with `double` now. This should fix #547

I also updated registered ops to use more general type - `realnumbertype`. Note, since I am not using default type for `T` ops compatibility test fails. Therefore, I have not updated yet `ops.pbtxt`.

I also would like to add `MaxPool3dGradGrad` (currently I added only forward declaration for it).

As this pull request is getting quite big, I would appreciate some comments.

**Comment by aam-at** [comment]
I updated 3d kernels. Cpu kernel supports only NHWC and gpu kernel supports both. Currently, python tests for 3d pooling check only NHWC because gpu kernels for pooling and pooling grad only work with this format.

Someone can correct me if I am wrong, but cudnn 5 supports both formats of tensors (NHWC and NCHW). However, both pooling 2d and 3d kernels with cudnn backend always convert tensors from NHWC to NCHW format.

**Comment by vrv** [comment]
This is really nice @aam-at: we're waiting for Benoit to do an initial review of this code (he's been on vacation the past week or so).

File: tensorflow/core/ops/nn_ops.cc
Comment (vrv) [inline tensorflow/core/ops/nn_ops.cc]: small note: this is doing validation that 'orig_input' and 'output' follow Pool3DShape's shape contract, but doesn't validate orig_output and 'grad'

I would recommend augmenting this registration with something like:

```
  .SetShapeFn([](InferenceContext* c) {
    TF_RETURN_IF_ERROR(shape_inference::Pool3DShape(c));
    ShapeHandle unused;
    // Validate 'grad' is the same shape as 'orig_input'
    TF_RETURN_IF_ERROR(c->Merge(c->input(0), c->input(2), &unused));
    // Validate 'orig_output' is same shape as 'output'
    TF_RETURN_IF_ERROR(c->Merge(c->input(1), c->output(0), &unused));
    return Status::OK();
  })
```

or something like that

File: tensorflow/core/ops/ops.pbtxt
Comment (vrv) [inline tensorflow/core/ops/ops.pbtxt]: Thanks for making these changes, though you don't need to add them in this PR -- they are automatically generated later from the op registrations.

File: tensorflow/core/ops/nn_ops.cc
Comment (aam-at) [inline tensorflow/core/ops/nn_ops.cc]: Thanks! I will add it in the next commit.

File: tensorflow/core/ops/ops.pbtxt
Comment (aam-at) [inline tensorflow/core/ops/ops.pbtxt]: Okay. I may drop these changes then. Also, as I noted before, the previous declaration of MaxPool3dGrad didn't use template attribute `T` for types of input and output. So ops compatibility tests will fail. I assume that the correct declaration is to use template attribute.

Comment (vrv) [inline tensorflow/core/ops/ops.pbtxt]: Oh, you probably need to read this: https://github.com/tensorflow/tensorflow/blob/754048a0453a04a761e112ae5d99c149eb9910dd/tensorflow/g3doc/how_tos/adding_an_op/index.md#backwards-compatibility

My suggestion would be to declare this as a new type Tinput with a default of DT_FLOAT to preserve backwards compatibility as in the suggestion in the link above.  It sucks to have to declare a new type for this, but I can't think of alternative that is backwards compatible.

**Comment by aam-at** [comment]
I dropped commit changing ops.pbtxt which required rebase.

File: tensorflow/core/ops/ops.pbtxt
Comment (aam-at) [inline tensorflow/core/ops/ops.pbtxt]: Yeah, that worked. I assume you meant something like this:
```c++
REGISTER_OP("MaxPool3DGrad")
    .Input("orig_input: Tinput")
    .Input("orig_output: Tinput")
    .Input("grad: T")
    .Output("output: T")
    .Attr("T: numbertype = DT_FLOAT")
    .Attr("Tinput: numbertype = DT_FLOAT");
```
Compatibility test passes now but I assume that original definition contained mistake. Do we want to preserve backward compatibility for this op?

**Comment by aam-at** [comment]
Any update on reviewing this?

**Comment by ahundt** [comment]
@vrv @benoitsteiner max pool grad grad definitely looks like some nice changes, would appreciate the review + merge!

**Comment by vrv** [comment]
I'll kick off a test anyway, but will wait for @benoitsteiner to review.

@tensorflow-jenkins test this please

**Comment by aam-at** [comment]
The build was broken due to the changes in the upstream (191658d54f90ac03c15b339326129cd52d1f56a3). It is a bit confusing that testing is performed after merging with upstream, though, quite reasonable.

I updated pooling with doubles to call new cudnn loading mechanism. So build should be successful now. @vrv Can you rerun tests again?

**Comment by vrv** [comment]
@tensorflow-jenkins test this please

(I don't know the status of our builds right now, so just trying anyway since it's the weekend and the load is low)

**Comment by vrv** [comment]
Looks like the sanity check is failing:

=== Sanity check step 3 of 6: do_buildifier (buildifier check) ===

Running do_buildifier on 157 files

tensorflow/core/kernels/BUILD # listsort unsafesort sort:tf_kernel_library.hdrs

buildifier took 0 s

FAIL: buildifier found errors and/or warnings in above BUILD files.
buildifier suggested the following changes:
2648d2647
<         "pooling_ops_common.h",
2649a2649
>         "pooling_ops_common.h",
Please fix manually or run buildifier <file> to auto-fix.

**Comment by aam-at** [comment]
Buildifier requires sorted list of headers and sources. So I fixed that. It's my
first pull request to tensorflow, so I am a bit confused with the requirements
for various tests. Thanks, @vrv!

**Comment by vrv** [comment]
@tensorflow-jenkins test this please

@gunan one thing that would be cool would be to have the sanity checks run automatically when users push their PRs -- but then we'd control the rest of the testing.

**Comment by vrv** [comment]
@aam-at this is a pretty epic first PR to TensorFlow, so thank you for your patience :)

**Comment by aam-at** [comment]
Tests are passing, except for macos. I don't have access to macos and I am not
sure why the build is failing. Thank you for your patience too, @vrv!

**Comment by vrv** [comment]
I think it's a infrastructure failure, the fact that the other tests pass suggests this is ready for review!

**Comment by gunan** [comment]
I think this can be an issue in build files. But bazel 0.4.3 has a bug in exception propagation, which make us see unhelpful messages. I will upgrade what I can access and rerun tests.

**Comment by gunan** [comment]
Jenkins, test this please

**Comment by aam-at** [comment]
All tests are passing now. Thanks for the help! I will wait for review now.

**Comment by vrv** [comment]
ping for @benoitsteiner for this week.

**Comment by ahundt** [comment]
@vrv @benoitsteiner Bumping this item since last post was 10 days ago. :-)

**Comment by aam-at** [comment]
@vrv @benoitsteiner any update on this?

**Comment by vrv** [comment]
@andydavis1 or @rmlarsen any takers? :)   I am swamped with a lot of other things :(

**Comment by jhseu** [comment]
@rmlarsen, do you have time to look at this change?

**Comment by rmlarsen** [comment]
I'll take a look a this sometime next week.

File: tensorflow/core/kernels/maxpooling_op.cc
Comment (rmlarsen) [inline tensorflow/core/kernels/maxpooling_op.cc]: Try:

OP_REQUIRES_OK(context, context->forward_input_or_allocate_output(
{2}, 0, tensor_out.shape(), &output));

File: tensorflow/core/kernels/maxpooling_op.cc
Comment (rmlarsen) [inline tensorflow/core/kernels/maxpooling_op.cc]: you don't need this. The OP_REQUIRES_OK macro tests the status and returns if not OK.

File: tensorflow/core/kernels/maxpooling_op.cc
Comment (rmlarsen) [inline tensorflow/core/kernels/maxpooling_op.cc]: you don't need this. The OP_REQUIRES_OK macro tests the status and returns if not OK.

File: tensorflow/core/kernels/maxpooling_op_gpu.cu.cc
Comment (rmlarsen) [inline tensorflow/core/kernels/maxpooling_op_gpu.cu.cc]: "mask" seems a little misleading. Isn't it really argmax within the window?

File: tensorflow/core/kernels/maxpooling_op_gpu.cu.cc
Comment (rmlarsen) [inline tensorflow/core/kernels/maxpooling_op_gpu.cu.cc]: I realize that you didn't start this, but could you please add a comment clarifying what "top" and "bottom" refer to? (I agree that it is preferably to use those terms compared to further overloading "input" and "output".)

Same everywhere.

**Review by rmlarsen — CHANGES_REQUESTED** [review]
First round of comments.

File: tensorflow/core/ops/nn_ops.cc
Comment (rmlarsen) [inline tensorflow/core/ops/nn_ops.cc]: Why allow separate types for input and output? The only existing registrations set them both to float:

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/pooling_ops_3d.cc#L287
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/pooling_ops_3d.cc#L490

File: tensorflow/core/ops/nn_ops.cc
Comment (rmlarsen) [inline tensorflow/core/ops/nn_ops.cc]: Update comment. Should be "gradients of gradients" or  "second-order" gradients".

File: tensorflow/python/kernel_tests/pooling_ops_test.py
Comment (rmlarsen) [inline tensorflow/python/kernel_tests/pooling_ops_test.py]: You can simplify this by using assertAllCloseAccordingToType().

File: tensorflow/python/kernel_tests/pooling_ops_test.py
Comment (rmlarsen) [inline tensorflow/python/kernel_tests/pooling_ops_test.py]: s/margin/tolerance/

File: tensorflow/python/kernel_tests/pooling_ops_test.py
Comment (rmlarsen) [inline tensorflow/python/kernel_tests/pooling_ops_test.py]: cut and paste error?

File: tensorflow/python/kernel_tests/pooling_ops_test.py
Comment (rmlarsen) [inline tensorflow/python/kernel_tests/pooling_ops_test.py]: This is broken. The test names are not unique and therefore only one of the possible tests will get run. This should set a test name that is a concatenation of all the test params. I also recommend using a function like 

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/kernel_tests/matmul_op_test.py#L35

to add the tests so this will not happen in the future.

File: tensorflow/stream_executor/cuda/cuda_dnn.cc
Comment (rmlarsen) [inline tensorflow/stream_executor/cuda/cuda_dnn.cc]: double?

File: tensorflow/stream_executor/cuda/cuda_dnn.cc
Comment (rmlarsen) [inline tensorflow/stream_executor/cuda/cuda_dnn.cc]: double?

**Review by rmlarsen — CHANGES_REQUESTED** [review]

File: tensorflow/stream_executor/cuda/cuda_dnn.cc
Comment (rmlarsen) [inline tensorflow/stream_executor/cuda/cuda_dnn.cc]: The cuDNN documentation states that:

"storage data type for alpha[0], beta[0] is float for HALF and SINGLE tensors, and double
for DOUBLE tensors. These parameters are passed using a host memory pointer"

File: tensorflow/core/kernels/maxpooling_op.cc
Comment (aam-at) [inline tensorflow/core/kernels/maxpooling_op.cc]: It seems that it was added only recently in the upstream. I will do rebase for that.

File: tensorflow/core/kernels/maxpooling_op.cc
Comment (aam-at) [inline tensorflow/core/kernels/maxpooling_op.cc]: Okay. I suspected that but previous implementation of GradOp had an explicit checks. I will remove unnecessary checks throughout `maxpooling` file then.

File: tensorflow/core/ops/nn_ops.cc
Comment (aam-at) [inline tensorflow/core/ops/nn_ops.cc]: It is related to compatibility tests. Previous registration uses `orig_input: float` and `grad: T`. If I try to generalize and set `orig_input: T`, the compatibility test will fail. It was previously discussed here - https://github.com/tensorflow/tensorflow/pull/6664#discussion_r96332231 and I followed @vrv suggestion.

File: tensorflow/core/ops/nn_ops.cc
Comment (aam-at) [inline tensorflow/core/ops/nn_ops.cc]: I updated doc to use your wording "second-order gradients". However, note that C++ code does not compute "final" gradients. GradGradOp will return gradient w.r.t. max pooling gradient. Second-order gradient will be correctly handled in backend gradient implementation, such as python.

File: tensorflow/core/kernels/maxpooling_op_gpu.cu.cc
Comment (aam-at) [inline tensorflow/core/kernels/maxpooling_op_gpu.cu.cc]: It has the same size as the output `N*C*Hout*Wout` and contains the corresponding index in the input for each element in the output.

File: tensorflow/python/kernel_tests/pooling_ops_test.py
Comment (aam-at) [inline tensorflow/python/kernel_tests/pooling_ops_test.py]: `GetShrunkInceptionMaxPoolShapes` returns unique name for each parameters test combination, so it should be okay.

**Comment by martinwicke** [comment]
@aam-at what's the status of this?

**Comment by rmlarsen** [comment]
@jlebar would you please review the stream executor changes in this PR?

**Comment by aam-at** [comment]
@martinwicke It implements MaxPoolingGradGrad kernels for CPU and GPU. Tests are passing on my machine.

File: tensorflow/core/kernels/maxpooling_op_gpu.cu.cc
Comment (rmlarsen) [inline tensorflow/core/kernels/maxpooling_op_gpu.cu.cc]: Exactly, so I would not name it mask, but "indices" or "argmax" or something like that.

**Comment by rmlarsen** [comment]
@aam-at Thanks for the updates. Could you please rebase to resolve the conflicts and address the remaining comments.

File: tensorflow/core/kernels/maxpooling_op_gpu.cu.cc
Comment (aam-at) [inline tensorflow/core/kernels/maxpooling_op_gpu.cu.cc]: Some kernels were borrowed from caffe, so we should rename `mask` there too. In addition, how should we rename functions like `MaxPoolBackwardNoMask`? In my opinion, the naming may be not the most intuitive (like `max_to_indices`) but it does not introduce too much confusion.

**Comment by aam-at** [comment]
@rmlarsen I have finished rebase. One commit was dropped, since `TensorFormat` handles now 5d tensors. Regarding `mask` parameter, I am not sure if we should rename this parameter because it can introduce further confusion:
1) some kernels were borrowed from caffe
2) not clear how to rename `NoMask` methods.

**Comment by martinwicke** [comment]
There are still (or again) conflicts. @rmlarsen if you can resolve the naming issues, we can move this PR along.

**Comment by rmlarsen** [comment]
@aam-at OK, let's stick with mask then.

**Comment by martinwicke** [comment]
(if this is the last sticking point, make sure to update your review, and I'll switch to just caring about the tests)

**Review by rmlarsen — APPROVED** [review]

**Comment by rmlarsen** [comment]
@aam-at please resolve the conflict so we can test this more thoroughly.

**Review by rmlarsen — CHANGES_REQUESTED** [review]
I would like to run our internal testing before approving this. This requires a clean rebase.

**Comment by aam-at** [comment]
@rmlarsen I did rebase again. Currently, there are no conflicts, so you can run the tests.

**Comment by rmlarsen** [comment]
@aam-at Thanks!

**Comment by rmlarsen** [comment]
@tensorflow-jenkins test this please

**Comment by rmlarsen** [comment]
@aam-at it looks like there are some broken tests (backwards compatibility and XLA pooling tests) that need attention.

File: tensorflow/core/ops/nn_ops.cc
Comment (rmlarsen) [inline tensorflow/core/ops/nn_ops.cc]: Do you need to keep the default value here?

File: tensorflow/core/ops/nn_ops.cc
Comment (rmlarsen) [inline tensorflow/core/ops/nn_ops.cc]: ditto?

File: tensorflow/core/ops/nn_ops.cc
Comment (aam-at) [inline tensorflow/core/ops/nn_ops.cc]: The problem was with `MaxPool3dGrad`. We discussed why it is necessary to have two types for input and output, so compatibility tests can pass. I guess during rebase I dropped those changes by accident.

File: tensorflow/core/ops/nn_ops.cc
Comment (aam-at) [inline tensorflow/core/ops/nn_ops.cc]: Fixed.

**Comment by aam-at** [comment]
@rmlarsen I fixed backward compatibility tests. Xla pooling ops test should be also fixed.

**Comment by vrv** [comment]
@tensorflow-jenkins test this please

**Review by rmlarsen — APPROVED** [review]

**Comment by rmlarsen** [comment]
@aam-at Excellent. Thank you for this large contribution!

**Comment by aam-at** [comment]
@vrv @rmlarsen Thank you for the help!

**Comment by rmlarsen** [comment]
@aam-at it appears that the two tests:

_testMaxPoolGradGradSamePadding2_1(data_format, use_gpu)

and 

_testMaxPoolGradGradSamePadding3_1(data_format, use_gpu)

are flaky when use_gpu is True, e.g.:

Traceback (most recent call last):
  File "/usr/grte/v4/k8-linux/lib/python2.7/unittest/case.py", line 331, in run
    testMethod()
  File "/build/work/0bfa20861d2fc543528291b5513e5ecb/google3/runfiles/google3/third_party/tensorflow/python/kernel_tests/pooling_ops_test.py", line 1293, in testMaxPoolGradGrad
    self._testMaxPoolGradGradSamePadding2_1(data_format, use_gpu)
  File "/build/work/0bfa20861d2fc543528291b5513e5ecb/google3/runfiles/google3/third_party/tensorflow/python/kernel_tests/pooling_ops_test.py", line 1258, in _testMaxPoolGradGradSamePadding2_1
    use_gpu=use_gpu)
  File "/build/work/0bfa20861d2fc543528291b5513e5ecb/google3/runfiles/google3/third_party/tensorflow/python/kernel_tests/pooling_ops_test.py", line 813, in _ConstructAndTestSecondGradient
    self.assertLess(err, err_tolerance)
  File "/usr/grte/v4/k8-linux/lib/python2.7/unittest/case.py", line 932, in assertLess
    self.fail(self._formatMessage(msg, standardMsg))
  File "/build/work/0bfa20861d2fc543528291b5513e5ecb/google3/runfiles/google3/testing/pybase/basetest.py", line 1205, in fail
    return super(TestCase, self).fail(self._formatMessage(prefix, msg))
  File "/usr/grte/v4/k8-linux/lib/python2.7/unittest/case.py", line 412, in fail
    raise self.failureException(msg)
AssertionError: 1.99932861328125 not less than 0.01

Perhaps there's a bug for padding is SAME? Do you have any idea what the underlying cause might be?

**Comment by rmlarsen** [comment]
The problem was caused by a race due to using buffer forwarding instead of allocation in MaxPoolingGradGradOp(). I'll submit a fix.

**Comment by aam-at** [comment]
@rmlarsen It was night in my timezone. Thanks for the fix. I added `forward_input_or_allocated_output` in c1b6aa4b6bbb2e80e228b3ecb9f2877f4e5dc26a. Shouldn't we replace buffer forwarding to allocating output for 3d pooling grad grad too? I am actually not sure why `forward_input_or_allocated_output` creates race conditions.

**Pull request merged**
